### PR TITLE
fixes wrong exit code being returned (#29)

### LINKED ISSUE #29: Returns wrong exit code on panic
I'm really liking criterion so far :heart: , I did encounter an issue where the wrong exit code is returned.

To reproduce:
- start from the criterion-rs readme example
- add `panic!("!!")` here:
```rust
fn criterion_benchmark(c: &mut Criterion) {
    panic!("!!");
    c.bench_function("fib 20", |b| b.iter(|| fibonacci(black_box(20))));
}
```
- run `cargo criterion`
- check the exit status of the previous command with `echo $?`
- this will return 0 but should be 101 because of the panic

What's strange is that the correct exit code is returned when the panic is moved to here:
```rust
fn fibonacci(n: u64) -> u64 {
    panic!("PANIC");
    match n {
        ...
```
The correct exit code is returned when using `cargo bench`.
Returning the wrong exit code is problematic in a CI context.

### PR DESCRIPTION
**Pull request by Anton-4**
- Fixed #29 by not returning immediately if the message is none but first checking the exit code of the spawned child process.
- Updated changelog.
- Clippy fixes (clippy 0.1.52 (9bc8c42 2021-05-09))

### CODE REVIEW

File: src/bench_target.rs [hunk 1]
```diff
@@ -4,7 +4,7 @@ use crate::report::{BenchmarkId, Report, ReportContext};
 use anyhow::{anyhow, Context, Result};
 use std::ffi::OsString;
 use std::net::TcpListener;
-use std::path::PathBuf;
+use std::path::{Path, PathBuf};
```
Comment (Anton-4) [inline src/bench_target.rs:7]: clippy

File: src/bench_target.rs [hunk 2]
```diff
@@ -27,7 +27,7 @@ impl BenchTarget {
     /// will block until the benchmark target terminates.
     pub fn execute(
         &self,
-        criterion_home: &PathBuf,
+        criterion_home: &Path,
```
Comment (Anton-4) [inline src/bench_target.rs:30]: clippy

File: src/connection.rs [hunk 3]
```diff
@@ -29,12 +29,12 @@ impl std::error::Error for ConnectionError {
 #[derive(Debug)]
 #[repr(u16)]
 enum ProtocolFormat {
-    CBOR = 1,
+    Cbor = 1,
```
Comment (Anton-4) [inline src/connection.rs:32]: clippy

File: src/connection.rs [hunk 4]
```diff
@@ -29,12 +29,12 @@ impl std::error::Error for ConnectionError {
...
 #[repr(u16)]
 enum ProtocolFormat {
-    CBOR = 1,
+    Cbor = 1,
 }
 impl ProtocolFormat {
     fn from_u16(format: u16) -> Result<Self, ConnectionError> {
         match format {
-            1 => Ok(ProtocolFormat::CBOR),
+            1 => Ok(ProtocolFormat::Cbor),
```
Comment (Anton-4) [inline src/connection.rs:37]: clippy

File: src/connection.rs [hunk 5]
```diff
@@ -257,10 +257,7 @@ pub enum SamplingMethod {
 }
 impl SamplingMethod {
     pub fn is_linear(&self) -> bool {
-        match self {
-            SamplingMethod::Linear => true,
-            _ => false,
-        }
+        matches!(self, SamplingMethod::Linear)
```
Comment (Anton-4) [inline src/connection.rs:260]: clippy

File: src/html/mod.rs [hunk 6]
```diff
@@ -24,7 +24,7 @@ use tinytemplate::TinyTemplate;
 
 const THUMBNAIL_SIZE: Option<Size> = Some(Size(450, 300));
 
-const COMMON_CSS: &'static str = include_str!("common.css");
+const COMMON_CSS: &str = include_str!("common.css");
```
Comment (Anton-4) [inline src/html/mod.rs:27]: clippy

File: src/message_formats/mod.rs [hunk 7]
```diff
@@ -4,8 +4,9 @@ use crate::config::{MessageFormat, SelfConfig};
 use self::json::JsonMessageReport;
 
 pub fn create_machine_report(self_config: &SelfConfig) -> Option<JsonMessageReport> {
-    match self_config.message_format {
-        Some(MessageFormat::Json) => Some(JsonMessageReport),
-        None => None,
+    if let Some(MessageFormat::Json) = self_config.message_format {
+        Some(JsonMessageReport)
+    } else {
+        None
```
Comment (Anton-4) [inline src/message_formats/mod.rs:10]: clippy

File: src/plot/gnuplot_backend/summary.rs [hunk 8]
```diff
@@ -37,16 +37,14 @@ pub fn line_comparison(
...
-
     figure.configure(Axis::LeftY, |a| {
         a.configure(Grid::Major, |g| g.show())
             .configure(Grid::Minor, |g| g.hide())
             .set(Label(format!("Average time ({})", unit)))
             .set(axis_scale.to_gnuplot())
     });
 
-    for (name, curve) in lines {
+    for (i, (name, curve)) in lines.iter().enumerate() {
```
Comment (Anton-4) [inline src/plot/gnuplot_backend/summary.rs:47]: clippy

File: src/stats/univariate/outliers/tukey.rs [hunk 9]
```diff
@@ -225,42 +225,27 @@ pub enum Label {
 impl Label {
     /// Checks if the data point has an "unusually" high value
     pub fn is_high(&self) -> bool {
-        match *self {
-            HighMild | HighSevere => true,
-            _ => false,
-        }
+        matches!(*self, HighMild | HighSevere)
```
Comment (Anton-4) [inline src/stats/univariate/outliers/tukey.rs:228]: clippy

**Comment by Anton-4** [comment]
I've found another case where the wrong exit code is being returned, I'm investigating it right now.

**Comment by Anton-4** [comment]
I think we should never return `Ok(())` when the message is `None` in `bench_target.rs` inside the `communicate` function.
In my case, it led to a panic resulting in a 0 exit code 90% of the time.

**Comment by Lemmih** [comment]
Thanks!

**Pull request merged**